ada-project-docs/wave_03.md

```diff
@@ -23,42 +24,24 @@ Things to note:
 - Possible sort fields:
   - Customers can be sorted by `name`, `registered_at` and `postal_code`
   - Videos can be sorted by `title` and `release_date`
-  - Overdue rentals can be sorted by `title`, `name`, `checkout_date` and `due_date`
 - If the client requests both sorting and pagination, pagination should be relative to the sorted order
 - Check out the [paginate method](https://flask-sqlalchemy.palletsprojects.com/en/2.x/api/#flask_sqlalchemy.BaseQuery.paginate)
 
-### More Endpoints: Inventory Management
-All these endpoints should support all 3 query parameters. All fields are sortable.
-
-#### `GET /rentals/overdue`
-List all customers with overdue videos
-
-Fields to return:
-- `video_id`
-- `title`
-- `customer_id`
-- `name`
-- `postal_code`
-- `checkout_date`
-- `due_date`
-
 
-#### `GET /videos/<id>/history`
-List customers that have checked out a copy of the video _in the past_
+#### Errors and Edge Cases to Check
 
-URI parameters:
-- `id`: Video identifier
+- The API should default to sorting by `id` if no `sort` parameter is specified
+- The API should default to sorting by `id` if a value other than `name`, `registered_at`, or `postal_code` is passed in for the parameter `sort`
+- The API should default to returning all customers in a single page if no number of per page responses is specified
+- The API should default to returning all customers in a single page if an invalid number of per page responses is specified
+- The API should default to returning the first page if no page is specified
+- The API should default to returning the first page if an invalid page is specified
 
-Fields to return:
-- `customer_id`
-- `name`
-- `postal_code`
-- `checkout_date`
-- `due_date`
+## Extra Endpoint: Inventory Management
+This endpoint should support all 3 query parameters. All fields are sortable.
 
-
-#### `GET /customers/<id>/history`
-List the videos a customer has checked out _in the past_
+### `GET /customers/<id>/history`
+List the videos a customer has checked out _in the past_. Current rentals should not be included.                                                                                  
```

Comment:
I could use other folk's input, but I'm leaning toward moving this endpoint to optional enhancements. Since project will be assigned in the afternoon on a Tuesday and is due before class the following Monday, there are really only 5 and a half days including the weekend of working time.

Reply:
What if we remove the query parameters for `/customers` and `/video/<id>/rentals` instead? That might remove some work that's a little tedious but is mostly repetition from what they've seen in Hello Books. They'll still have practice with `/customers/<id>/rentals`

I do like that the history endpoint requires them to rethink how they structure their model and some other endpoints, but I understand that might also make it the best candidate for cutting out.

Reply:
I like your points. I don't have data on how long students historically had to complete the project, so I'm still concerned about time pressures if they need to make significant design changes. 

I feel like pagination is really common in modern websites, so it could be an interesting and helpful topic to have a little more understanding of how it can be done, even though it's based in query parameters which they'll have had some practice in. We also don't really touch on how to build up a query that has multiple sorting or filtering options, which I think could be useful to know. If we cut the endpoint and the concern is that this wave is too trivial, would it be worth adding a couple filter options that should be stackable?

Reply:
Hm yeah, it looks like they did have 17 days to complete it. The two intervening weeks were company presentations. 
 
@tgoslee Do you have any insights about how C16 did with time management on this project?

The changes I chose to make to accommodate the history endpoint were to add a boolean `checked_in` to the Rental model, and instead of deleting a rental from the db when a video is checked in, just toggle the . `/customers/<id>/rentals` should then also be modified so we only.

One thing we could do is put the `/customers/<id>/history` endpoint _before_ the rentals endpoint so they don't have to retroactively make changes to their models and endpoints. My only thought there is they will prioritize that over query parameters stuff if they are running low on time, and they may find it confusing to create endpoints about customer rental history before they do `rentals/check-in` and `rentals/check-out`

Reply:
Hmm... what are the learning goals we are looking to meet with the `/customers/<id>/history` endpoint that we don't get from the others? I don't love this, but since Accelerate students have so much less working time than prior cohorts, if there isn't something specific students should get from it that the other endpoints won't provide, my inclination is still to cut it over rearranging the order in the waves.

Reply:
This project was paired and they did have a lot of time so there actually weren't complaints about time with this project.